From a401b623aa41198661045c9de84906faf1806b84 Mon Sep 17 00:00:00 2001
From: Austin Schuh <austin.linux@gmail.com>
Date: Sat, 1 Feb 2025 01:11:51 -0800
Subject: [PATCH] Allow reducing reported UVC compressed bandwidth

This fixes out of bandwidth problems without needing to fix the device.

Upstream-Status: Inappropriate [device hack]
Signed-off-by: Austin Schuh <austin.linux@gmail.com>
---

--- a/drivers/media/usb/uvc/uvc_video.c
+++ b/drivers/media/usb/uvc/uvc_video.c
@@ -25,6 +25,9 @@
 #include <media/v4l2-common.h>
 
 #include "uvcvideo.h"
+
+static unsigned int bandwidth_quirk_divisor = 1;
+module_param(bandwidth_quirk_divisor, uint, 0444);
 
 /* ------------------------------------------------------------------------
  * UVC Controls
@@ -258,6 +261,17 @@
 		bandwidth = max_t(u32, bandwidth, 1024);
 
 		ctrl->dwMaxPayloadTransferSize = bandwidth;
+	} else if ((format->flags & UVC_FMT_FLAG_COMPRESSED) &&
+		   (stream->dev->quirks & UVC_QUIRK_FIX_BANDWIDTH) &&
+		   stream->intf->num_altsetting > 1) {
+		u32 prior = ctrl->dwMaxPayloadTransferSize;
+
+		/* Guard against zero divisors and rounding small payloads to zero. */
+		ctrl->dwMaxPayloadTransferSize = max_t(u32, 1,
+			prior / max_t(unsigned int, bandwidth_quirk_divisor, 1));
+		uvc_dbg(stream->dev, VIDEO,
+			"Reducing compressed bandwidth from %u B/frame to %u B/frame\n",
+			prior, ctrl->dwMaxPayloadTransferSize);
 	}
 
 	if (stream->intf->num_altsetting > 1 &&
